### PR TITLE
Support GOPATH with multiple entries when importing the coverage (Fix #308)

### LINKED ISSUE #308: InvalidPathException when importing coverage results on windows machine
Hy. Trying to import covergae results from windows machine, but getting:
`java.nio.file.InvalidPathException: Illegal char <:> at index 20: C:\Users\zalent\go;C:\work\go\src\<trimmed_full_path>`
On linux works fine. maybe because i have two paths in my gopath env variable?

### PR DESCRIPTION
**Pull request by alban-auzeill**
Extract GOPATH path resolution mechanism so test report sensor could use it

### CODE REVIEW

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    assertThat(linuxContext.concat("", "paul")).isEqualTo("paul");
+
+    assertThat(windowsContext.concat("C:\\Users", "paul")).isEqualTo("C:\\Users\\paul");
+    assertThat(windowsContext.concat("C:\\Users\\", "paul")).isEqualTo("C:\\Users\\paul");
+    assertThat(windowsContext.concat("C:\\", "paul")).isEqualTo("C:\\paul");
+    assertThat(windowsContext.concat("", "paul")).isEqualTo("paul");
+  }
+
+  @Test
+  void source_directories() {
```
Comment (vilchik-elena) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java]: `multiple_source_directories_in_gopath`

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    GoPathContext linuxContext = new GoPathContext('/', ":", "/home/paul/go:/home/luc/go");
+    assertThat(linuxContext.goSrcPathList).containsExactly("/home/paul/go/src", "/home/luc/go/src");
+
+    GoPathContext windowsContext = new GoPathContext('\\', ";", "C:\\Users\\paul\\go;C:\\Users\\luc\\go");
+    assertThat(windowsContext.goSrcPathList).containsExactly("C:\\Users\\paul\\go\\src", "C:\\Users\\luc\\go\\src");
+  }
+
+  @Test
+  void resolve_using_go_path() throws IOException {
```
Comment (vilchik-elena) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java]: split this test in several

Comment (vilchik-elena) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java]: add test with 2 valid path in gopath

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    context = new GoPathContext(File.separatorChar, File.pathSeparator, goPath);
+    expected = "file.go";
+    assertThat(context.resolveUsingGoPath("file.go")).isEqualTo(expected);
+  }
+
+  @Test
+  void resolve_absolute_path() throws IOException {
+    GoPathContext linuxContext = new GoPathContext('/', ":", "/home/paul/go");
+    assertThat(linuxContext.resolveUsingGoPath("_/my-app/my-app.go")).isEqualTo("/my-app/my-app.go");
+    assertThat(linuxContext.resolveUsingGoPath("my-app/my-app.go")).isEqualTo("/home/paul/go/src/my-app/my-app.go");
```
Comment (vilchik-elena) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java]: line can be removed

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+    GoPathContext windowsContext = new GoPathContext('\\', ";", "C:\\Users\\paul\\go");
+    assertThat(windowsContext.resolveUsingGoPath("_\\C_\\my-app\\my-app.go")).isEqualTo("C:\\my-app\\my-app.go");
+    assertThat(windowsContext.resolveUsingGoPath("my-app\\my-app.go")).isEqualTo("C:\\Users\\paul\\go\\src\\my-app\\my-app.go");
```
Comment (vilchik-elena) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java]: same

File: sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoPathContext.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  String concat(String parentPath, String childPath) {
+    if (parentPath.isEmpty() || parentPath.charAt(parentPath.length() - 1) == fileSeparator) {
+      return parentPath + childPath;
+    }
+    return parentPath + fileSeparator + childPath;
+  }
+
+  String resolveUsingGoPath(String filePath) {
```
Comment (vilchik-elena) [inline sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoPathContext.java]: how about just `resolve`?

File: sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoPathContext.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  private Optional<String> prefixByFirstValidGoPath(String filePath) {
+    return goSrcPathList.stream()
+      .map(goPath -> concat(goPath, filePath))
+      .filter(path -> new File(path).exists())
+      .findFirst();
+  }
+
+  private String prefixByFirstGoPath(String filePath) {
```
Comment (vilchik-elena) [inline sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoPathContext.java:101]: can you add comment explaining this hack

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class GoPathContextTest {
+
+  @Test
+  void concat() {
```
Comment (vilchik-elena) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java]: split test

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+    assertThat(windowsContext.concat("C:\\Users", "paul")).isEqualTo("C:\\Users\\paul");
+    assertThat(windowsContext.concat("C:\\Users\\", "paul")).isEqualTo("C:\\Users\\paul");
+    assertThat(windowsContext.concat("C:\\", "paul")).isEqualTo("C:\\paul");
+    assertThat(windowsContext.concat("", "paul")).isEqualTo("paul");
+  }
+
+  @Test
+  void source_directories() {
+    GoPathContext linuxContext = new GoPathContext('/', ":", "/home/paul/go:/home/luc/go");
```
Comment (vilchik-elena) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoPathContextTest.java]: how about having path separator at the end? `"/home/paul/go:/home/luc/go:"`

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoPathContext.java (hunk 6 above)
Comment (alban-auzeill) [reply sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoPathContext.java:101]: added on the resolve(...) method

**Comment by alban-auzeill** [comment]
@vilchik-elena please review my last commit

File: sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoPathContext.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+  String concat(String parentPath, String childPath) {
+    if (parentPath.isEmpty() || parentPath.charAt(parentPath.length() - 1) == fileSeparator) {
+      return parentPath + childPath;
+    }
+    return parentPath + fileSeparator + childPath;
+  }
+
+  /**
+   * Try to resolve the absolute path of the given filePath. If filePath is relative,
```
Comment (vilchik-elena) [inline sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoPathContext.java]: @alban-auzeill You forgot to mention about normalization of already absolute path

File: sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoPathContext.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+   * try to append the first GOPATH entry where this file exists, otherwise return
+   * a non-existing absolute path using the first GOPATH entry (or just filePath itself
+   * if GOPATH is empty).
+   * See GoCoverageReport#findInputFile to understand why a non-existing file is useful.
```
Comment (vilchik-elena) [inline sonar-go-plugin/src/main/java/org/sonar/go/plugin/GoPathContext.java]: @alban-auzeill this "See" is mean:) there is no doc for `GoCoverageReport#findInputFile`. You could also use `{@link..}`

**Review by vilchik-elena — APPROVED** [review]

**Pull request merged**